### ISSUE TITLE
kubseal --raw and types

### ISSUE THREAD

**Issue by remmeier** [bitnami-labs/sealed-secrets#372]
I attempted to seal a secret for a dockerconfigjson-typed secret:

```
apiVersion: bitnami.com/v1alpha1
kind: SealedSecret
metadata:
  annotations:
    sealedsecrets.bitnami.com/cluster-wide: "true"
  creationTimestamp: null
  name: registry-cred
spec:
  encryptedData:
    .dockerconfigjson: ...
  template:
    type: kubernetes.io/dockerconfigjson
```

from the controller I got:

```
type: 'Warning' reason: 'ErrUpdateFailed' Secret "sb4b-registry-cred" is invalid: data[.dockerconfigjson
]: Invalid value: "<secret contents redacted>": invalid character 'e' looking for beginning of value
```

After trying around for a while I noticed that the secret can be unsealed if I remove `type: kubernetes.io/dockerconfigjson`. It also worked when I do not make use of  `--raw` but rather seal the complete secret.

So I assume the type is used as criteria for the sealing? I did not see a command line parameter to pass the type with raw. Or I may missed something?

**Comment by mkmik** [comment]
can you paste paste the whole process of sealing your secret that you used? possibly using real structure in your .dockerconfigjson content (without the actual secrets of course). From the error message it seems that the value that has been decrypted cannot be interpreted as a valid json

**Comment by ka-zsolt** [comment]
@mkmik  I have the exact same problem with the newest version of 0.17.1

```
export DOCKER_REGISTRY=my.docker.registry
export DOCKER_USERNAME=dockeruser
export DOCKER_PASSWORD=dockerpassword

export DOCKER_AUTH=$(echo -n $DOCKER_USERNAME:$DOCKER_PASSWORD|base64 -w 0)
export DOCKER_AUTHS=$(echo -n "{\"auths\": {\"$DOCKER_REGISTRY\": {\"auth\": \"$DOCKER_AUTH\" }}}" | base64 -w 0)

echo -n $DOCKER_AUTHS | base64 -d
{"auths": {"my.docker.registry": {"auth": "ZG9ja2VydXNlcjpkb2NrZXJwYXNzd29yZA==" }}}%

echo -n ZG9ja2VydXNlcjpkb2NrZXJwYXNzd29yZA== | base64 -d
dockeruser:dockerpassword

export DOCKER_AUTHS_SEALED=$(echo -n $DOCKER_AUTHS | kubeseal --raw --scope cluster-wide  --cert publickey.crt --from-file=/dev/stdin)

cat <<EOF | kubectl apply -f -
apiVersion: bitnami.com/v1alpha1
kind: SealedSecret
metadata:
  annotations:
    sealedsecrets.bitnami.com/cluster-wide: "true"
  name: docker-registry-credentials-secret
spec:
  encryptedData:
    .dockerconfigjson: $DOCKER_AUTHS_SEALED
  template:
    metadata:
      annotations:
        sealedsecrets.bitnami.com/cluster-wide: "true"
        sealedsecrets.bitnami.com/managed: "true"
      name: docker-registry-credentials-secret
    type: kubernetes.io/dockerconfigjson
EOF
    
sealedsecret.bitnami.com/docker-registry-credentials-secret created

kubectl describe sealedsecrets.bitnami.com docker-registry-credentials-secret
...
Status:
  Conditions:
    Last Transition Time:  2021-12-16T18:14:45Z
    Last Update Time:      2021-12-16T18:14:45Z
    Message:               Secret "docker-registry-credentials-secret" is invalid: data[.dockerconfigjson]: Invalid value: "<secret contents redacted>": invalid character 'e' looking for beginning of value
    Status:                False
    Type:                  Synced
  Observed Generation:     1
Events:
  Type     Reason           Age                 From            Message
  ----     ------           ----                ----            -------
  Warning  ErrUpdateFailed  13s (x12 over 18s)  sealed-secrets  Secret "docker-registry-credentials-secret" is invalid: data[.dockerconfigjson]: Invalid value: "<secret contents redacted>": invalid character 'e' looking for beginning of value

```

**Comment by mkmik** [comment]
Can you make a report with some fake secret data so you can feel comfortable sharing the actual encrypted value?

**Comment by ka-zsolt** [comment]
Sure, but if you run this - copy paste the whole and run - you should get the same results, give sealed-secrets is installed in kube-system namespace. I've just tried it on a fresh minikube. Tell me if you need anything else!

```
export DOCKER_REGISTRY=my.docker.registry
export DOCKER_USERNAME=dockeruser
export DOCKER_PASSWORD=dockerpassword

export DOCKER_AUTH=$(echo -n $DOCKER_USERNAME:$DOCKER_PASSWORD|base64 -w 0)
export DOCKER_AUTHS=$(echo -n "{\"auths\": {\"$DOCKER_REGISTRY\": {\"auth\": \"$DOCKER_AUTH\" }}}" | base64 -w 0)

export DOCKER_AUTHS_SEALED=$(echo -n $DOCKER_AUTHS | kubeseal --raw --scope cluster-wide  --from-file=/dev/stdin)

cat <<EOF | kubectl apply -f -
apiVersion: bitnami.com/v1alpha1
kind: SealedSecret
metadata:
  annotations:
    sealedsecrets.bitnami.com/cluster-wide: "true"
  name: docker-registry-credentials-secret
spec:
  encryptedData:
    .dockerconfigjson: $DOCKER_AUTHS_SEALED
  template:
    metadata:
      annotations:
        sealedsecrets.bitnami.com/cluster-wide: "true"
        sealedsecrets.bitnami.com/managed: "true"
      name: docker-registry-credentials-secret
    type: kubernetes.io/dockerconfigjson
EOF
 
```
output of 
kubectl describe sealedsecrets.bitnami.com docker-registry-credentials-secret

```
Name:         docker-registry-credentials-secret
Namespace:    default
Labels:       <none>
Annotations:  sealedsecrets.bitnami.com/cluster-wide: true
API Version:  bitnami.com/v1alpha1
Kind:         SealedSecret
Metadata:
  Creation Timestamp:  2021-12-16T18:51:55Z
  Generation:          1
  Managed Fields:
    API Version:  bitnami.com/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:status:
    Manager:      controller
    Operation:    Update
    Subresource:  status
    Time:         2021-12-16T18:51:55Z
    API Version:  bitnami.com/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:kubectl.kubernetes.io/last-applied-configuration:
          f:sealedsecrets.bitnami.com/cluster-wide:
      f:spec:
        .:
        f:encryptedData:
        f:template:
    Manager:         kubectl-client-side-apply
    Operation:       Update
    Time:            2021-12-16T18:51:55Z
  Resource Version:  761
  UID:               1877415c-3553-42e7-9450-a81f275b41fc
Spec:
  Encrypted Data:
    .dockerconfigjson:  AgBEEjK3WA2ASXiP3x4zciyoA0FpAfJ1sQYcMmXBHGzURarub+WPH/6SKB+1IKzx3xnmONjBuXYgHDIX2dOxuFrF8Vc6iDGhbUPsatwEMWvO1pVDBQQqYPA/aFT119Rx5RISYcUjt+OtUSVofTnBT+gX6EOcLsKU//9zVWb3d9EZzvhLvQXoUDm2TCrP9iWPT3u1iIqGyhmeMGFwxuS7OqZBfqoKDfVljrW/ghRg0NNbVQ9OSRMaS0Dnr1F0qduq0bccPeVV7Iu5RwVf3TkXHa7qvECzBT6BeR4LkPBjtnB6d9K0rHHfYbuKPqH1pHhVpEBrVKJte9I0ZorOECD4p+wMtlM+NQaIKTUWCX13LRNpNdhvorKhHy289V+XEpSipc7C/sLFX99PrtDX1e4+fYdIrwaq5MZ014EBEHgMRQ8nxR3nNROoWQfP8LDTvwPA0SH2194EVjDLNgYyrHAkBCm75Tn/kpC1zFTLMqNHymJacdUiK6LPLEZUj5zPeZnW4WHFC6XflNXxLdmgU3hrRGP3Ks9I7+fRD8ykUe8JCcu3Huj4o3c/dQuWlFPM6P/3BfvwBmpJScESR1niytEaNthR1CTKmyb/mLfCG2FfIk5DoTvt+hYY3JnqUkNrIOd7V/AwlXwk4ueFmtlWMfNG4L3TO8PAmuOXW8IrJ2h2MvvrMMegla7p8JQx2oyj8VJboQzwgU6JpLcrRQvLrcpY/gNwuFp6RlHP79tP6vFMfny1Sw0DIGjDlsX3LUg5fX7voftgJ71G4lS7R2xNDFigFkx/wf+SMPyk/8U5Me+S+az8ls0L3U9923FUpLeKU9idDNhjerUwm6vjHapCxBVdUPRn
  Template:
    Metadata:
      Annotations:
        sealedsecrets.bitnami.com/cluster-wide:  true
        sealedsecrets.bitnami.com/managed:       true
      Name:                                      docker-registry-credentials-secret
    Type:                                        kubernetes.io/dockerconfigjson
Status:
  Conditions:
    Last Transition Time:  2021-12-16T18:51:55Z
    Last Update Time:      2021-12-16T18:51:55Z
    Message:               Secret "docker-registry-credentials-secret" is invalid: data[.dockerconfigjson]: Invalid value: "<secret contents redacted>": invalid character 'e' looking for beginning of value
    Status:                False
    Type:                  Synced
  Observed Generation:     1
Events:
  Type     Reason           Age                   From            Message
  ----     ------           ----                  ----            -------
  Warning  ErrUpdateFailed  107s (x12 over 112s)  sealed-secrets  Secret "docker-registry-credentials-secret" is invalid: data[.dockerconfigjson]: Invalid value: "<secret contents redacted>": invalid character 'e' looking for beginning of value
```

**Comment by ka-zsolt** [comment]
@mkmik Is my previous comment sufficient to replicate the problem on your side?

**Comment by mkmik** [comment]
@ka-zsolt yes

The error message is:

```
invalid character 'e' looking for beginning of value
```

This is k8s attempting to parse a JSON value but instead getting a base64's json string.

```console
$ export DOCKER_AUTHS=$(echo -n "{\"auths\": {\"$DOCKER_REGISTRY\": {\"auth\": \"$DOCKER_AUTH\" }}}" | base64 -w 0)
$ echo -n $DOCKER_AUTHS
eyJhdXRocyI6IHsibXkuZG9ja2VyLnJlZ2lzdHJ5IjogeyJhdXRoIjogIlpHOWphMlZ5ZFhObGNqcGtiMk5yWlhKd1lYTnpkMjl5WkE9PSIgfX19
```

K8s is attempting to parse the JSON because you told it that the type of the secret is `kubernetes.io/dockerconfigjson`

---

The fix is to simply avoid the spurious base64 encoding step:

```console
$ export DOCKER_AUTHS=$(echo -n "{\"auths\": {\"$DOCKER_REGISTRY\": {\"auth\": \"$DOCKER_AUTH\" }}}")
$ export DOCKER_AUTHS_SEALED=$(echo -n $DOCKER_AUTHS | kubeseal --raw --scope cluster-wide  --from-file=/dev/stdin)
....
```

or alternatively:

```console
$ export DOCKER_AUTH=$(echo -n $DOCKER_USERNAME:$DOCKER_PASSWORD|base64 -w 0)
$ export DOCKER_AUTHS=$(echo -n "{\"auths\": {\"$DOCKER_REGISTRY\": {\"auth\": \"$DOCKER_AUTH\" }}}" )
$ cat <<EOF | kubectl apply -f -
apiVersion: bitnami.com/v1alpha1
kind: SealedSecret
metadata:
  annotations:
    sealedsecrets.bitnami.com/cluster-wide: "true"
  name: docker-registry-credentials-secret
spec:
  encryptedData:
    .dockerconfigjson: $(echo -n $DOCKER_AUTHS | kubeseal --raw --scope cluster-wide)
  template:
    metadata:
      annotations:
        sealedsecrets.bitnami.com/cluster-wide: "true"
        sealedsecrets.bitnami.com/managed: "true"
      name: docker-registry-credentials-secret
    type: kubernetes.io/dockerconfigjson
EOF
```

---

You need to think about the "base64" aspect of k8s secrets as a way to encode arbitrary binary data in a JSON/YAML file.
I.e. the base64 encoding of the secrets is just a "transport" mechanism between you and kubernetes.

The input to `kubeseal` is the secret *as is* without any need to encode it as base64. `kubeseal` will encrypt whatever byte stream it receives and produce an encrypted output that is suitable for inserting in a SealedSecret JSON/YAML resource (which means it has to be ascii, i.e. it will in turn be encoded as base64)

**Comment by ka-zsolt** [comment]
Thank you for your detailed answer, it is working now!

My misconception was that in a vanilla k8s secret when I use "stringData:" instead of the encoded "data:" the actual data will be kept unencoded in the secret (how and where it is actually stored is a different topic). To understand it I created a secret with stringData and did a `kubectl get secret mysecret -o yaml` and I expected to get back this:
```
stringData:
  mykey: "myvalue"
```
But instead I got back data encoded:
```
data:
  mykey: bXl2YWx1ZQ==
```

Thanks again to help clarify this for me!
